CAIPs/caip-211.md

```diff
@@ -0,0 +1,275 @@
+---
+caip: 211
+title: JSON-RPC Authority Negotiation
+author: Pedro Gomes (@pedrouid), Hassan Malik (@hmalik88)
+discussions-to: https://github.com/ChainAgnostic/CAIPs/pull/207
+status: Draft
+type: Standard
+created: 2023-02-02
+updated: 2023-02-02
+requires: [2, 10, 25, 171]
+---
+
+## Simple Summary
+
+CAIP-211 extends CAIP-25 to allow callers and respondents to anchor feature
+discovery in specific RPC semantics and to request the respondent route requests
+to specific RPC endpoints.
+
+## Abstract
+
+CAIP-211 defines additional properties that enable progressive usage of CAIP-25
+to layer custom or local RPC semantics and/or routing onto a session. Since
+CAIP-25 respondents ignore unknown properties, respondents that conform to
+CAIP-25 but not to CAIP-211 should be carefully considered and accomodated by
+implementers. 
+
+## Motivation
+
+While some core methods and notifications are foundational to entire namespaces
+and thus almost universally defined out-of-band (meaning all callers and
+respondents agree to them already), others are specific to chains or even to
+subsets of wallets and dapps on a given chain.  This requires scope objects in
+CAIP-25 requests to negotation semantic anchors and/or network routing
+**before** authorization can occur in the confidence that both parties agree to
+the syntax and semantics of a given method or notification, which may only be
+available on certain RPC endpoints.
+
+The concept of a custom RPC endpoint or definition presumes the existence of
+"default" or universal endpoints and definitions, which can be hard to anchor in
+explicit specifications unless a namespace profile of this CAIP has been
+published. (This may be unnecessary in the case of younger namespaces without
+customization built in at the per-chain or per-dapp layer). An empty array of
+`rpcEndpoints` or `rpcDocuments` sent in either direction should be interpreted
+differently than the absence of the array.  Namely, either array being present
+but empty in a request signals that a caller is requesting that custom endpoints
+or  definitions NOT be considered in CAIP-25 authorizations. Present but empty
+arrays in a response confirms this behavior, while its absence can be taken to
+mean the wallet either does not support custom endpoints/definitions or the user
+has opted out of that degree of trust, which for privacy/fingerprinting reasons,
+should not be handled by distinct codepaths (See Privacy Considerations).
+
+## Specification
+
+Two properties are added to the scope objects requested in `optionalScopes` and
```

Comment:
These properties need to be allowed in `requiredScopes` as well, doesn't make sense to have them in one and not the other.

Reply:
That's fair - I hadn't really been considering that use case, but there will probably be dapps that you can't even (or wouldn't even want to) connect your wallet to without mutually-supported custom RPC, right? I thought the examples should only have them in `optionalScopes` because that will be the most common pattern.

Reply:
fixed [here](https://github.com/ChainAgnostic/CAIPs/pull/211/commits/f91abb189234dc4ae0733dfcfb266396ebbfc231)